zkevm-circuits/src/evm_circuit/op_execution/utils/memory_gadgets.rs

```diff
@@ -0,0 +1,231 @@
+        )
+        .get_lower_128() as Address
+    }
+}
+
+/// The sum of bytes of the address that are unused for most calculations on the address
+pub(crate) mod address_high {
+    use super::super::super::Word;
+    use super::super::NUM_ADDRESS_BYTES_USED;
+    use crate::evm_circuit::op_execution::utils;
+    use halo2::{arithmetic::FieldExt, plonk::Expression};
+
+    pub(crate) fn expr<F: FieldExt>(address: &Word<F>) -> Expression<F> {
+        utils::sum::expr(&address.cells[NUM_ADDRESS_BYTES_USED..32].to_vec())
+    }
+
+    pub(crate) fn value<F: FieldExt>(address: [u8; 32]) -> F {
+        utils::sum::value::<F>(&address[NUM_ADDRESS_BYTES_USED..32].to_vec())
+    }
+}
+
+/// If any of the MSBs are non-zero we're always out of gas
+pub(crate) fn require_address_in_range<F: FieldExt>(
+    cb: &mut ConstraintBuilder<F>,
+    address: &Word<F>,
+) {
+    cb.require_zero(address_high::expr(address));
+}
+
+/// Calculates the memory size required for a memory access at the specified address.
+/// `memory_size = ceil(address/32) = floor((address + 31) / 32)`
+#[derive(Clone, Debug)]
+pub(crate) struct MemorySizeGadget<F> {
+    memory_size: ConstantDivisionGadget<F, MAX_MEMORY_SIZE_IN_BYTES>,
+}
+
+impl<F: FieldExt> MemorySizeGadget<F> {
+    pub const NUM_CELLS: usize =
+        ConstantDivisionGadget::<F, MAX_MEMORY_SIZE_IN_BYTES>::NUM_CELLS;
+    pub const NUM_WORDS: usize =
+        ConstantDivisionGadget::<F, MAX_MEMORY_SIZE_IN_BYTES>::NUM_WORDS;
+
+    pub(crate) fn construct(alloc: &mut CaseAllocation<F>) -> Self {
+        Self {
+            memory_size: ConstantDivisionGadget::construct(alloc, 32),
+        }
+    }
+
+    pub(crate) fn constraints(
+        &self,
+        cb: &mut ConstraintBuilder<F>,
+        address: Expression<F>,
+    ) -> Expression<F> {
+        let (quotient, _remainder) =
+            self.memory_size.constraints(cb, address + 31.expr());
+        quotient
+    }
+
+    pub(crate) fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        address: Address,
+    ) -> Result<MemorySize, Error> {
+        let (quotient, _) =
+            self.memory_size
+                .assign(region, offset, (address as u128) + 31)?;
+        Ok(quotient as MemorySize)
+    }
+}
+
+/// Returns (new memory size, memory gas cost) for a memory access.
+/// If the memory needs to be expanded this will result in an extra gas cost.
+/// This gas cost is the difference between the next and current memory costs:
+/// `memory_cost = Gmem * memory_size + floor(memory_size * memory_size / 512)`
+#[derive(Clone, Debug)]
+pub(crate) struct MemoryExpansionGadget<F> {
+    address_memory_size: MemorySizeGadget<F>,
+    next_memory_size: MaxGadget<F, MAX_MEMORY_SIZE_IN_BYTES>,
+    curr_quad_memory_cost:
+        ConstantDivisionGadget<F, { MAX_MEMORY_SIZE_IN_BYTES * 2 }>,
+    next_quad_memory_cost:
+        ConstantDivisionGadget<F, { MAX_MEMORY_SIZE_IN_BYTES * 2 }>,
```

Comment:
I see, used for the OOG case makes sense to me. So the `u128` is then needed.

> I currently made the max quad cost size configurable so that we can indeed change this to 8 for the success case, but allow it to be set bigger to support all possible input values for the out-of-gas case. Does that make sense to you?

Yeah, make sense!

> Yes, for sure. My assumption was that this would be done externally of the op gadgets because practically all op codes require this.

Yes, or naively we can also have a `gas_gadget.rs` to help that.